internal/utils/azure/azure.go

```diff
@@ -53,94 +75,119 @@ func NewAzureAPI() (*AzureAPI, error) {
 // AzureDenyAssignmentsClient is a facade over the Azure deny assignments client. Exists to make our
 // code easier to test (it handles paging).
 type AzureDenyAssignmentsClient struct {
+	ctx    context.Context
 	client *armauthorization.DenyAssignmentsClient
 }
 
 // NewAzureDenyAssignmentsClient creates a new AzureDenyAssignmentsClient (our facade client) from a
 // client from the Azure SDK.
-func NewAzureDenyAssignmentsClient(azClient *armauthorization.DenyAssignmentsClient) *AzureDenyAssignmentsClient {
+func NewAzureDenyAssignmentsClient(ctx context.Context, azClient *armauthorization.DenyAssignmentsClient) *AzureDenyAssignmentsClient {
 	return &AzureDenyAssignmentsClient{
+		ctx:    ctx,
 		client: azClient,
 	}
 }
 
 // GetDenyAssignmentsForScope gets all the deny assignments matching a scope and an optional filter.
 func (c *AzureDenyAssignmentsClient) GetDenyAssignmentsForScope(scope string, filter *string) ([]*armauthorization.DenyAssignment, error) {
+	var denyAssignments []*armauthorization.DenyAssignment
 	pager := c.client.NewListForScopePager(scope, &armauthorization.DenyAssignmentsClientListForScopeOptions{
 		Filter: filter,
 	})
 
-	var denyAssignments []*armauthorization.DenyAssignment
-	for pager.More() {
-		nextResult, err := pager.NextPage(context.TODO())
-		if err != nil {
-			return nil, fmt.Errorf("failed to get next page of results: %w", err)
-		}
-		if nextResult.Value != nil {
-			denyAssignments = append(denyAssignments, nextResult.Value...)
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		for pager.More() {
+			nextResult, err := pager.NextPage(c.ctx)
+			if err != nil {
+				ch <- fmt.Errorf("failed to get next page of results: %w", err)
+			}
+			if nextResult.Value != nil {
+				denyAssignments = append(denyAssignments, nextResult.Value...)
+			}
 		}
+		ch <- nil
+	}()
+
+	select {
+	case err := <-ch:
+		return denyAssignments, err
+	case <-c.ctx.Done():
+		return denyAssignments, fmt.Errorf("context cancelled")
```

Comment:
shouldnt we also add a case for nil being put on the channel and then just `return denyAssignments, nil`?